src/audio/pipewire.c

```diff
@@ -0,0 +1,277 @@
+    struct spa_ringbuffer rb;    // a thread-safe ring buffer implementation which uses atomic operations  to guarantee some semblance of thread safety, therefore it doesn't require a mutex
+    char *sample_buffer;         // the heap storage memory which will be backing the atomic ring buffer implementation, it's on the heap because this struct would be transfered across callback functions a lot, and such a large buffer could cause a stack overflow on some hardware architectures
+    uint32_t stride;             // the amount of bytes per frame. This is in state because it's computed dynamically, according to each format specifyer
+    char *sink_name;             // the name of the sink, as given by speech dispatcher
+} module_state;
+
+// pipewire on process callback
+// this function would be called each time the server requests data from us
+static void on_process(void *userdata)
+{
+    module_state *state = (module_state *)userdata;
+    // a lot of the code below has been taken from the playing a tone pipewire tutorial, see the pipewire documentation for more details
+
+    struct pw_buffer *b;
+    struct spa_buffer *buf;
+    uint32_t *destination_memory, read_index, available_for_loading, must_load_with_silence;
+    int32_t available_samples, number_of_frames;
+    if ((b = pw_stream_dequeue_buffer(state->stream)) == NULL)
+    {
+        pw_log_warn("out of buffers: %m");
+        return;
+    }
+    buf = b->buffer;
+    // pipewire's buffers are mapped directly from the server to the client
+    // As such, if the memory we're given doesn't exist, we have to abort imediately, because doing otherwise most certainly would lead to undefined behaviour
+    if ((destination_memory = buf->datas[0].data) == NULL)
+        return;
+    // in the process callback, we take the samples from the ring buffer we used in the pipewire_play function and send it to pipewire for playback
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_samples = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    number_of_frames = buf->datas[0].maxsize / state->stride; // how many frames of audio are in this chunk, aka how many channel_count touples are available this call to on_process. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases
+    if (b->requested > 0)
+    {
+        number_of_frames = SPA_MIN(number_of_frames, b->requested);
+    }
```

Comment:
Apparently `b->requested` was introduced in version 0.3.49 only, not available in CI. So you need to surround this piece with `#if PW_CHECK_VERSION(0, 3, 49)`